python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1500,11 +1500,30 @@ def should_backfill_atomic_asset_partitions_unit(
                 f" {parent_partitions_result.required_but_nonexistent_parents_partitions}"
             )
 
+        asset_partitions_to_request_map: Dict[AssetKey, Set[str]] = {}
+        for asset_partition in asset_partitions_to_request:
+            asset_partitions_to_request_map.setdefault(asset_partition.asset_key, set()).add(
+                asset_partition.partition_key
+            )
         for parent in parent_partitions_result.parent_partitions:
             can_run_with_parent = (
                 parent in asset_partitions_to_request
                 and asset_graph.have_same_partitioning(parent.asset_key, candidate.asset_key)
-                and parent.partition_key == candidate.partition_key
+                and (
+                    (
+                        parent.partition_key == candidate.partition_key
+                        and not asset_graph.get_backfill_policy(parent.asset_key)
+                    )
+                    # candidate shares a partition key that is already being requested by the parent
+                    # and both candidate and parent have backfill policies and max_partitions_per_run is not 1
+                    or (
+                        asset_graph.get_backfill_policy(parent.asset_key)
+                        and asset_graph.get_backfill_policy(parent.asset_key).max_partitions_per_run
```

Comment:
I think this logic would be easier to parse if it was factored out in the following way:

```python
for parent in parent_partitions_result.parent_partitions:
    parent_backfill_policy = asset_graph.get_backfill_policy(parent.asset_key)
    # checks if this parent partition mapping has a backfill policy which would allow
    # partition mappings which are not one-one with the child to be executed in a way
    # that respects the dependencies
    has_partition_mapping_safe_backfill_policy = (
        parent_backfill_policy == asset_graph.get_backfill_policy(candidate.asset_key)
        and parent_backfill_policy is not None
        and (
            # if multiple runs for a single asset would be created this tick, we can't be confident that
           # the partition dependencies will be respected between those two separate runs
            parent_backfill_policy.max_partitions_per_run >= 
            len(asset_partitions_to_request_map[parent.asset_key])
        )
    )
    can_run_with_parent = (
        ...
        and (
            parent.partition_key == candidate.partition_key or (
                has_partition_mapping_safe_backfill_policy
                and candidate.partition_key in asset_partitions_to_request_map[parent.asset_key]
            )
        )
        ...
    )

```

Reply:
Note that this slightly changes the logic to not just look for max_partitions_per_run == 1, but rather the general case of situations where the tick would produce multiple runs for a given asset key (e.g. you have a max_partitions_per_run of 2, but 3 partitions were requested)